### PR TITLE
[Potential Flow] Fixing Compressible Perturbation Element

### PR DESCRIPTION
**Pull request by inigolcanalejo**
**Description**
Adding the correct lhs and rhs for clamped and lower wake elements.

**Changelog**
- Added cpp tests for all rhs and lhs cases: clamped/unclamped and normal/wake
- Added elemental sensitivities cpp tests for all cases: clamped/unclamped and normal/wake
- Fixed lhs for clamped normal and wake elements
- Fixed lhs and rhs for lower wake elements
- Added function ComputePerturbedVelocityLowerElement
- Added function ComputeVacuumVelocitySquared and corresponding tests
- Removed unused deprecated ComputeDensity and ComputeDensityDerivative functions

### CODE REVIEW

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 1]
```diff
@@ -200,7 +200,11 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateOnInt
     }
     else if (rVariable == DENSITY)
     {
-        rValues[0] = ComputeDensity(rCurrentProcessInfo);
+        const array_1d<double, Dim> velocity = PotentialFlowUtilities::ComputePerturbedVelocity<Dim,NumNodes>(*this, rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]:
```suggestion
        const array_1d<double, Dim>& r_velocity = PotentialFlowUtilities::ComputePerturbedVelocity<Dim,NumNodes>(*this, rCurrentProcessInfo);
```

no need for a copy here since it is const

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 2]
```diff
@@ -409,24 +413,13 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateLeftH
...
-    const double density = ComputeDensity(rCurrentProcessInfo);
-    const double DrhoDu2 = ComputeDensityDerivative(density, rCurrentProcessInfo);
-
-    // Computing local velocity
-    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
-    array_1d<double, Dim> velocity = PotentialFlowUtilities::ComputeVelocity<Dim,NumNodes>(*this);
-    for (unsigned int i = 0; i < Dim; i++){
-        velocity[i] += free_stream_velocity[i];
-    }
+    const array_1d<double, Dim> velocity = PotentialFlowUtilities::ComputePerturbedVelocity<Dim,NumNodes>(*this, rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]: same

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 3]
```diff
@@ -442,16 +435,14 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateRight
...
-    const double density = ComputeDensity(rCurrentProcessInfo);
+    const CompressiblePerturbationPotentialFlowElement& r_this = *this;
 
-    // Computing local velocity
-    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
-    array_1d<double, Dim> velocity = PotentialFlowUtilities::ComputeVelocity<Dim,NumNodes>(*this);
-    for (unsigned int i = 0; i < Dim; i++){
-        velocity[i] += free_stream_velocity[i];
-    }
+    array_1d<double, Dim> velocity = PotentialFlowUtilities::ComputePerturbedVelocity<Dim,NumNodes>(r_this, rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:440]: same

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 4]
```diff
@@ -470,32 +461,30 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateLeftH
     GeometryUtils::CalculateGeometryData(GetGeometry(), data.DN_DX, data.N, data.vol);
     GetWakeDistances(data.distances);
 
-    const double density = ComputeDensity(rCurrentProcessInfo);
-    const double DrhoDu2 = ComputeDensityDerivative(density, rCurrentProcessInfo);
+    // Compute upper and lower velocities
+    const array_1d<double, Dim> upper_velocity = PotentialFlowUtilities::ComputePerturbedVelocity<Dim,NumNodes>(*this, rCurrentProcessInfo);
+    const array_1d<double, Dim> lower_velocity = PotentialFlowUtilities::ComputePerturbedVelocityLowerElement<Dim,NumNodes>(*this, rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]: same

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 5]
```diff
@@ -515,21 +504,21 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateRight
     GeometryUtils::CalculateGeometryData(r_geometry, data.DN_DX, data.N, data.vol);
     GetWakeDistances(data.distances);
 
-    const double density = ComputeDensity(rCurrentProcessInfo);
+    // Compute upper and lower velocities
+    const array_1d<double, Dim> upper_velocity = PotentialFlowUtilities::ComputePerturbedVelocity<Dim,NumNodes>(*this, rCurrentProcessInfo);
+    const array_1d<double, Dim> lower_velocity = PotentialFlowUtilities::ComputePerturbedVelocityLowerElement<Dim,NumNodes>(*this, rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]: same

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 6]
```diff
@@ -589,34 +619,46 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateLeftH
         Points, data.DN_DX, data.distances, Volumes, GPShapeFunctionValues,
         PartitionsSign, GradientsValue, NEnriched);
 
-    const double density = ComputeDensity(rCurrentProcessInfo);
-    const double DrhoDu2 = ComputeDensityDerivative(density, rCurrentProcessInfo);
+    // Compute upper and lower velocities
+    const array_1d<double, Dim> upper_velocity = PotentialFlowUtilities::ComputePerturbedVelocity<Dim,NumNodes>(*this, rCurrentProcessInfo);
+    const array_1d<double, Dim> lower_velocity = PotentialFlowUtilities::ComputePerturbedVelocityLowerElement<Dim,NumNodes>(*this, rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]: same

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp [hunk 7]
```diff
@@ -67,40 +146,63 @@ KRATOS_TEST_CASE_IN_SUITE(CompressiblePerturbationPotentialFlowElementRHS, Compr
...
+    AssignPotentialsToNormalCompressiblePerturbationElement(pElement, potential);
 
     // Compute RHS
     Vector RHS = ZeroVector(3);
 
     const ProcessInfo& r_current_process_info = model_part.GetProcessInfo();
     pElement->CalculateRightHandSide(RHS, r_current_process_info);
 
-    std::vector<double> reference{146.2643261263345,-122.1426284341492,-24.12169769218525};
+    std::vector<double> reference{131.4361747323354,-113.768439084114,-17.66773564822145};
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp:158]:
```suggestion
    std::array<double, 3> reference{131.4361747323354,-113.768439084114,-17.66773564822145};
```

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp [hunk 8]
```diff
@@ -67,40 +146,63 @@ KRATOS_TEST_CASE_IN_SUITE(CompressiblePerturbationPotentialFlowElementRHS, Compr
...
+    std::array<double, 3> potential{1.0, 220.0, 250.0};
+    AssignPotentialsToNormalCompressiblePerturbationElement(pElement, potential);
+
+    // Compute RHS
+    Vector RHS = ZeroVector(3);
+
+    const ProcessInfo& r_current_process_info = model_part.GetProcessInfo();
+    pElement->CalculateRightHandSide(RHS, r_current_process_info);
+
+    std::vector<double> reference{205.3219372530133,-190.7662916232804,-14.55564562973297};
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp:179]:
```suggestion
    std::array<double, 3> reference{205.3219372530133,-190.7662916232804,-14.55564562973297};
```

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp [hunk 9]
```diff
@@ -205,7 +459,36 @@ KRATOS_TEST_CASE_IN_SUITE(WakeCompressiblePerturbationPotentialFlowElementRHS, C
...
+    const std::array<double, 6> potential{1.0, 151.0, 190.0, 6.0, 165.0, 195.0};
+    AssignPotentialsToWakeCompressiblePerturbationElement(pElement, distances, potential);
+
+    // Compute RHS and LHS
+    Vector RHS = ZeroVector(6);
+
+    const ProcessInfo& r_current_process_info = model_part.GetProcessInfo();
+    pElement->CalculateRightHandSide(RHS, r_current_process_info);
+
+    std::vector<double> reference{171.8439523046275,11.025,-4.730584829663217,5.5125,-161.6550003638144,-10.91673422229973};
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp:487]:
```suggestion
    std::array<double, 6> reference{171.8439523046275,11.025,-4.730584829663217,5.5125,-161.6550003638144,-10.91673422229973};
```

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp [hunk 10]
```diff
@@ -246,5 +560,185 @@ KRATOS_TEST_CASE_IN_SUITE(WakeCompressiblePerturbationPotentialFlowElementLHS, C
...
+    ComputeWakeElementalSensitivities(model_part, LHS_finite_diference, LHS_analytical, potential);
+
+    PrintTestElementInfo(model_part);
+
+    for (unsigned int i = 0; i < LHS_finite_diference.size1(); i++) {
+        for (unsigned int j = 0; j < LHS_finite_diference.size2(); j++) {
+            if(std::abs(LHS_finite_diference(i,j)-LHS_analytical(i,j)) > 1e-10){
+                KRATOS_WATCH(i)
+                KRATOS_WATCH(j)
+                KRATOS_WATCH(std::abs(LHS_finite_diference(i,j)-LHS_analytical(i,j)))
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp]: intentional?

**Review by philbucher — COMMENTED** [review]
minor, mostly accessing the array3 with reference (I might have missed some), if they are anyway const

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 11]
```diff
@@ -409,24 +413,13 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateLeftH
...
+    BoundedMatrix<double, NumNodes, NumNodes> lhs = ZeroMatrix(NumNodes,NumNodes);
 
-    noalias(rLeftHandSideMatrix) +=
-        data.vol * density * prod(data.DN_DX, trans(data.DN_DX));
-    noalias(rLeftHandSideMatrix) += data.vol * 2 * DrhoDu2 * outer_prod(DNV, trans(DNV));
+    CalculateLeftHandSideContribution(lhs, rCurrentProcessInfo, velocity, data);
 
-    const BoundedMatrix<double, NumNodes, NumNodes> rLaplacianMatrix =
-        data.vol * density * prod(data.DN_DX, trans(data.DN_DX));
+    noalias(rLeftHandSideMatrix) += lhs;
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]: Why summing the contribution instead of assigning it?

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 12]
```diff
@@ -553,6 +542,47 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateRight
...
+    // Compute lhs
+    const BoundedVector<double, NumNodes> DNV = prod(rData.DN_DX, rVelocity);
+
+    rLhs_total = rData.vol * density * prod(rData.DN_DX, trans(rData.DN_DX));
+
+    const double local_velocity_squared = inner_prod(rVelocity, rVelocity);
+
+    const double max_velocity_squared = PotentialFlowUtilities::ComputeMaximumVelocitySquared<Dim, NumNodes>(rCurrentProcessInfo);
+    if (local_velocity_squared < max_velocity_squared){
+        rLhs_total += rData.vol * 2 * DrhoDu2 * outer_prod(DNV, trans(DNV));
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: Does this improve the system a lot when approaching M=1?

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp (hunk 7 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp:158]: > minor, mostly accessing the array3 with reference (I might have missed some), if they are anyway const

Thanks @philbucher, I'll go through your suggestions

**Comment by inigolcanalejo** [comment]
> minor, mostly accessing the array3 with reference (I might have missed some), if they are anyway const

Thanks @philbucher, I'll go through your suggestions

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp (hunk 7 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp:158]: This doesn't compile:

kratos/includes/exception.h:87:16: error: no match for ‘operator<<’ (operand types are ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ and ‘const std::array<double, 3ul>’)
         buffer << rValue;

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp (hunk 11 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]: My bad

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp (hunk 12 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: Yes, this is basically what we were missing for the clamped elements and it actually allows to have M >1

**Comment by inigolcanalejo** [comment]
@philbucher and @marcnunezc thank you for your comments!

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp [hunk 13]
```diff
@@ -762,77 +816,28 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::ComputePotenti
     array_1d<double, NumNodes> distances;
     GetWakeDistances(distances);
 
+    auto r_geometry = GetGeometry();
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp]:
```suggestion
    auto& r_geometry = GetGeometry();
```

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp (hunk 12 above)
Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: Cool!
Btw, any reason to use the velocity squared over the mach number to check this condition?

I see this depends on the user define MACH_SQUARED_LIMIT. Wouldn''t it be better to have everything depending on a single parameter? (e.g MACH_LIMT only, and derive the rest from this)

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: Good point, I think I see what you mean but I don't see the way of using the mach number to check this condition. The free stream velocity and mach number define the vacuum velocity, at which the local mach number becomes infinite and above which it is not defined. This is why we don't allow the mach number to go above certain value and the clamped elements do not "see" their real local mach number. This is the reason why I chose to compare the velocities instead of the mach numbers. We can discuss it in the next meeting, maybe I'm missing something.

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: What I mean is that when computing the MaximumVelocitSquared, we are using:

https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L158

Which is a user defined quantity. 

On one hand, since we also have the `MACH_LIMIT` as a user defined paramter, I wonder if we can remove the squared one, and do here ` MACH_LIMIT**2`. Maybe this is coming from literature?

On the other hand, my question was then: why not comparing the `local_mach_number < MACH_LIMIT`? Wouldn't it be the same? meaning that we dont allow local_mach numbers higher than this value.

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: > On one hand, since we also have the `MACH_LIMIT` as a user defined paramter, I wonder if we can remove the squared one, and do here ` MACH_LIMIT**2`. Maybe this is coming from literature?

I agree we should only have one parameter. I would actually keep  ` MACH_LIMIT**2` because it is the term appearing in all of the formulas. This is the reason why in the transonic solver the limit is usually also set in the literature to  ` MACH_LIMIT**2 = 3.0` instead of using ` MACH_LIMIT`.  I didn't remove ` MACH_LIMIT` because it is still used in the non-perturbated compressible element. Do you have anything against keeping ` MACH_LIMIT**2` and removing  ` MACH_LIMIT`? Shall I adapt the non-perturbated compressible element to this?



> On the other hand, my question was then: why not comparing the `local_mach_number < MACH_LIMIT`? Wouldn't it be the same? meaning that we dont allow local_mach numbers higher than this value.

It is not the same. To make the comparison `local_mach_number < MACH_LIMIT` we would need to compute the  `local_mach_number` and this is not always possible, as we would get nan for high velocities. Let me try to explain what I mean with some plots. This is how the local mach number looks like vs the velocity magnitude:

Screenshot from 2021-03-08 11-16-05

As you can see, when approaching the velocity 807 m/s  the local mach number tends to infinity and beyond 807 m/s  it is not defined (807 m/s is the vacuum speed where the isentropic formulas that we are using are already way beyond their limits of application). In 2D such large velocities are usually not reached, but in 3D it can easily happen near the wing-tip trailing edge (it already happened to me when running some 3d examples).

This happens because the local mach number squared becomes negative beyond 807 m/s (which is of course physically impossible):

Screenshot from 2021-03-08 11-24-03

And this happens because the speed of sound squared becomes negative beyond the vacuum speed:

Screenshot from 2021-03-08 11-27-52

This is by the way the reason why we need to introduce a limit on the mach number on the first place. The velocity magnitude squared, however, is always well defined:

Screenshot from 2021-03-08 11-32-30

Which is the reason why the velocity is one used for the checks instead of the mach number (including inside the computation of the local mach number itself):

https://github.com/KratosMultiphysics/Kratos/blob/293aab0852cdc8c642e010a1e5e5ca84b7f40555/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L179-L188

Having said this, I think we need to improve some stuff:

- Remove one of the limits as you suggested.
- ComputeMaximumVelocitySquared only once and save it in the processinfo (Ill open another PR for this)
- What we do in [ComputeLocalMachNumber](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L502-L513) is very dangerous because it will give nan in 3D. So I would remove:  [ComputeLocalMachNumber](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L502-L513) and [ComputeLocalSpeedOfSound](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L404-L427) (I can also do this in another PR and adapt the non-perturbated elements to the safe way of computing the local mach number)
- Also at some point I would remove the non-perturbated elements and keep only the perturbated ones to clean up a bit (but I think we still have all of the embedded elements and the adjoints depending on them so we can do this later)

What do you think? Does this make sense to you? Maybe you have other comments? (I am pretty sure  we can improve and refactor the application a lot and I would like to do this once we have the 3D elements running)

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: Sorry for the above long answer, I didn't know how to make it shorter :sweat_smile:

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: Thanks for your detailed answer! It makes sense. 

> - Remove one of the limits as you suggested.

My only concern is a "practical" one, as it more intuitive to choose (as a user) a flat mach_number rather than a squared value. If in the code we keep the MACH_LIMIT**2 check, do you think it makes sense to make the user input the mach_numer limit (not squared). Or would you avoid this?

> - ComputeMaximumVelocitySquared only once and save it in the processinfo (Ill open another PR for this)
>- What we do in ComputeLocalMachNumber is very dangerous because it will give nan in 3D. So I would remove: ComputeLocalMachNumber and ComputeLocalSpeedOfSound (I can also do this in another PR and adapt the non-perturbated elements to the safe way of computing the local mach number)

Ok to me adressing this in a different PR

> Also at some point I would remove the non-perturbated elements and keep only the perturbated ones to clean up a bit (but I think we still have all of the embedded elements and the adjoints depending on them so we can do this later)

I agree with this but as you say the embedded/adjoint are depending on the un-perturbed elements and it will require some work to adapt and make sure eveyrthing is working correclty.

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp [hunk 14]
```diff
@@ -67,40 +146,63 @@ KRATOS_TEST_CASE_IN_SUITE(CompressiblePerturbationPotentialFlowElementRHS, Compr
...
 
-/** Checks the CompressiblePerturbationPotentialFlowElement.
- * Checks the LHS computation.
- */
+KRATOS_TEST_CASE_IN_SUITE(CompressiblePerturbationPotentialFlowElementRHSClamping, CompressiblePotentialApplicationFastSuite) {
+    Model this_model;
+    ModelPart& model_part = this_model.CreateModelPart("Main", 3);
+
+    GenerateCompressiblePerturbationElement(model_part);
+    Element::Pointer pElement = model_part.pGetElement(1);
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp:168]:
```suggestion
    Element::Pointer p_element = model_part.pGetElement(1);
```
Minor comment:
Just to say that this does not follow the style-guide (it is wrong in mostly all tests). Probably better to do a separate PR cleaning up everything

**Review by marcnunezc — APPROVED** [review]

File: applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp (hunk 12 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp:568]: > My only concern is a "practical" one, as it more intuitive to choose (as a user) a flat mach_number rather than a squared value. If in the code we keep the MACH_LIMIT**2 check, do you think it makes sense to make the user input the mach_numer limit (not squared). Or would you avoid this?

I know what you mean and honestly I don't know what is the best solution. From one point of view, I agree with you that thinking in terms of a flat mach_number is more intuitive and also in the purely subsonic compressible element the limit is usually set directly on the mach number. For example Eller proposes to use M=0.94, although for the Lock cases I was usually using M=0.99. For the transonic cases however since usually MACH_LIMIT**2=3.0 is used, I find this easier rather than having to input MACH_LIMIT=sqrt(3.0)=1.73205... Which for me also raises the question, which mach limit (or mach limit squared) to set as a default since we have two different values depending whether we are computing subsonic or transonic... Shall we move this setting to the [solver were we set up the element](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py#L46-L101)?

I would suggest that for now we remove MACH_LIMIT**2 from the input as you say and we move this setting to the set up of the element since it is element dependent, would you agree with this?

> Ok to me adressing this in a different PR

Ok, I'll try to do it this week

> I agree with this but as you say the embedded/adjoint are depending on the un-perturbed elements and it will require some work to adapt and make sure eveyrthing is working correclty.

Makes sense to me :+1:

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp (hunk 14 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_compressible_perturbation_potential_element.cpp:168]: Ok! Ill do it in an another PR as you say!
